vpr/src/base/vpr_constraints.cpp

```diff
@@ -58,6 +60,70 @@ PartitionRegion VprConstraints::get_partition_pr(PartitionId part_id) {
     return pr;
 }
 
+void VprConstraints::add_route_constraint(RouteConstraint rc) {
+    route_constraints_.insert({rc.get_net_name(), rc});
+    return;
+}
+
+RouteConstraint VprConstraints::get_route_constraint_by_net_name(std::string net_name) {
+    RouteConstraint rc;
+    auto const& rc_itr = route_constraints_.find(net_name);
+    if (rc_itr == route_constraints_.end()) {
+        // try regexp
+        bool found_thru_regex = false;
+        for (auto constraint : route_constraints_) {
+            if (std::regex_match(net_name, std::regex(constraint.first))) {
+                {
+                    rc = constraint.second;
+
+                    // mark as invalid so write constraint function will not write constraint
+                    // of regexpr name
+                    // instead a matched constraint is inserted in
+                    constraint.second.set_is_valid(false);
+                    rc.set_net_name(net_name);
+                    rc.set_is_valid(true);
+                    route_constraints_.insert({net_name, rc});
+
+                    found_thru_regex = true;
+                    break;
+                }
+            }
+        }
+        if (!found_thru_regex) {
+            rc.set_net_name("INVALID");
+            rc.set_net_type("INVALID");
+            rc.set_route_model("INVALID");
+            rc.set_is_valid(false);
+        }
+    } else {
+        rc = rc_itr->second;
+    }
+    return rc;
+}
+
+RouteConstraint VprConstraints::get_route_constraint_by_idx(std::size_t idx) const {
+    RouteConstraint rc;
+    if ((route_constraints_.size() == 0) || (idx > route_constraints_.size() - 1)) {
+        rc.set_net_name("INVALID");
```

Comment:
It seems to me that the ``set_net_name()``, ``set_net_type()`` and ``set_route_model()`` are redundant here. If the ``is_valid()`` return false, it does not need to be considered.
Maybe I miss some details, open to discussion.